lib/Launcher.js

```diff
@@ -412,6 +604,52 @@ function getWSEndpoint(browserURL) {
   });
 }
 
+/**
+ * @param {ChromeLauncher|FirefoxLauncher} launcher
+ *
+ * @return {{executablePath: string, missingText: ?string}}
+ */
+function resolveExecutablePath(launcher) {
+  // puppeteer-core doesn't take into account PUPPETEER_* env variables.
+  if (!launcher._isPuppeteerCore) {
+    const executablePath = process.env.PUPPETEER_EXECUTABLE_PATH || process.env.npm_config_puppeteer_executable_path || process.env.npm_package_config_puppeteer_executable_path;
+    if (executablePath) {
+      const missingText = !fs.existsSync(executablePath) ? 'Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: ' + executablePath : null;
+      return { executablePath, missingText };
+    }
+  }
+  const browserFetcher = new BrowserFetcher(launcher._projectRoot);
+  if (!launcher._isPuppeteerCore) {
+    const revision = process.env['PUPPETEER_CHROMIUM_REVISION'];
+    if (revision) {
+      const revisionInfo = browserFetcher.revisionInfo(revision);
+      const missingText = !revisionInfo.local ? 'Tried to use PUPPETEER_CHROMIUM_REVISION env variable to launch browser but did not find executable at: ' + revisionInfo.executablePath : null;
+      return {executablePath: revisionInfo.executablePath, missingText};
+    }
+  }
+  const revisionInfo = browserFetcher.revisionInfo(launcher._preferredRevision);
+  const missingText = !revisionInfo.local ? `Browser is not downloaded. Run "npm install" or "yarn install"` : null;
+  return {executablePath: revisionInfo.executablePath, missingText};
+}
+
+/**
+ * @param {string} product
+ * @param {string} projectRoot
+ * @param {string} preferredRevision
+ * @param {boolean} isPuppeteerCore
+ * @return {ChromeLauncher|FirefoxLauncher}
+ */
+function Launcher(product, projectRoot, preferredRevision, isPuppeteerCore) {
```

Comment:
Let's rename this `launch` now that it's no longer a class

Reply:
I'm happy to, but it implies a load of type annotation changes (e.g. `Launcher.BrowserOptions`), and I'm not sure that's what you want. 

I left it as `Launcher`, but do let me know if you still want `launcher`.